### PR TITLE
Adding lesson on the cascade

### LINKED ISSUE #4780: The Cascade [Learn CSS]
- [Draft](https://docs.google.com/document/d/1qxZydMVapxlKviMFBl0trH3ejg5i18l3NlMp2XwSMl4/edit)
- [PR](https://github.com/GoogleChrome/web.dev/pull/4984)
- [Preview](https://deploy-preview-4984--web-dev-staging.netlify.app/learn/css/the-cascade/)

## To do

- [ ] Self-assessment for position
- [ ] Self-assessment for origin
- [ ] Una to review
- [x] Adam to review

### PR DESCRIPTION
**Pull request by rachelandrew**
Fixes #4780 

See the linked issue for outstanding to do items on this lesson.

### CODE REVIEW

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+<figure class="w-figure">
+{% Codepen {
+  user: 'web-dot-dev',
+  id: 'xxgbPMP'
+} %}
+</figure>
+
+{% Aside %}
+This approach of declaring the same property twice works because browsers ignore values they don't understand.
```
Comment (una) [inline src/site/content/en/learn/css/the-cascade/index.md:124]: Maybe add a callout how: "Unlike other programming languages, CSS will not throw an error or break your program when it detects a line it cannot parse. Instead, that line is ignored."

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+such as a browser extension that allows a user to write their own custom CSS for a webpage.
+3. **Authored CSS**. The CSS that you author.
+4. **Authored `!important`**. Any `!important` that you add to your authored declarations.
+5. **Local user styles `!important`**. Any `!important` that come from the operating system level,
+or browser extension level CSS.
+6. **User agent `!important`**. Any `!important` that are defined in the default CSS,
+provided by the browser.
+
+<figure class="w-figure">
+{% Img src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/rjZMVAC5lNATzCErbyeI.png",
```
Comment (una) [inline src/site/content/en/learn/css/the-cascade/index.md:232]: Do we have an SVG of this?

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+its expected behaviour is to change visual state.
+
+## Using DevTools to find out why some CSS is not applying
+
+Browser DevTools will typically show all CSS that could match an element,
+with those which are not being used crossed out.
+
+<figure class="w-figure">
+{% Img src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/LA0lZUYkptGrwrDsLEe1.png",
+alt="An image of browser DevTools with overwritten CSS crossed out", width="800", height="446" %}
```
Comment (una) [inline src/site/content/en/learn/css/the-cascade/index.md:266]: Love the idea behind this image. I'd love to see the text be bigger. There's a lot of whitespace and it's hard to read the DevTools bit.

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+  user: 'web-dot-dev',
+  id: 'GRrgMOm'
+} %}
+</figure>
+
+Understanding the cascade algorithm helps you understand how the browser resolves conflicts like this.
+The cascade algorithm is split into 4 distinct stages.
+
+1. **Position and order of appearance**: the order of which your CSS rules appear
+2. **Specificity**: an algorithm which determines which CSS selector is the most specific
```
[the remark this answers is not in the document]
Comment (rachelandrew) [reply src/site/content/en/learn/css/the-cascade/index.md:38]: is most targeted doesn't make sense, maybe "has the strongest match"?

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+CSS from a browser extension, or your authored CSS
+4. **Importance**: some CSS rules are treated with more importance than others,
+especially with the `!important` rule type
+
+## Position and order of appearance
+
+The order in which your CSS rules appear and how they appear is taken into consideration by the cascade,
+ while it calculates conflict resolution.
+
+The demo right at the start of this lesson is the most straightforward example of position.
```
[the remark this answers is not in the document]
Comment (rachelandrew) [reply src/site/content/en/learn/css/the-cascade/index.md:49]: I would say start is more clear than top, personally.

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+There are two rules that have selectors of identical specificity,
+so the last one to be declared won.
+
+Styles can come from various sources on an HTML page,
+such as a `<link>` tag,
+an embedded `<style>` tag,
+ and inline CSS as defined in an element's `style` attribute.
+
+If you have a `<link />` that includes CSS at the top of your HTML page,
```
[the remark this answers is not in the document]
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:58]: Yeh that's cool. Definitely don't have a strong opinion on this.

File: src/site/content/en/learn/css/the-cascade/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:38]: Strongest match sounds good to me

File: src/site/content/en/learn/css/the-cascade/index.md (hunk 2 above)
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:232]: Yeh absolutely. I'm just putting bitmaps up for diagrams while we get things ready to go, so I don't have to spend time messing around converting text elements to paths etc to much. We should add switching these things to SVG as a pre-flight check.

File: src/site/content/en/learn/css/the-cascade/index.md (hunk 3 above)
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:266]: Yeh I'll do a tighter crop on that one

File: src/site/content/en/learn/css/the-cascade/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:38]: I'd prefer the reference it having the highest score, but we haven't introduced that yet at this point

File: src/site/content/en/learn/css/the-cascade/index.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+1. **Position and order of appearance**: the order of which your CSS rules appear
+2. **Specificity**: an algorithm which determines which CSS selector is the most specific
+3. **Origin**: the order of when CSS appears and where it comes from, whether that is a browser style,
+CSS from a browser extension, or your authored CSS
+4. **Importance**: some CSS rules are treated with more importance than others,
+especially with the `!important` rule type
+
+## Position and order of appearance
+
+The order in which your CSS rules appear and how they appear is taken into consideration by the cascade,
```
[the remark this answers is not in the document]
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:46]: nah good catch

File: src/site/content/en/learn/css/the-cascade/index.md (hunk 1 above)
Comment (hankchizljaw) [reply src/site/content/en/learn/css/the-cascade/index.md:124]: Yeh I like that

**Comment by hankchizljaw** [comment]
Right, tweaked some bits and bobs, so fancy having a quick scan @rachelandrew? 

I think we should add a stage in the content spreadsheet that signals that the PR is ready to rock but we just need to go through and stick final versions of diagrams in SVG format where appropriate too. That’s the sort of thing I can do in batches.

**Review by hankchizljaw — APPROVED** [review]

**Review by argyleink — APPROVED** [review]
Very nice how this ended up 👍🏻 

My only feedback is the final image, devtools rules crossed out, seemed like it could use a caption. But, since other's have reviewed and didn't seem to need one, i feel it's not important.

**Comment by rachelandrew** [comment]
> Right, tweaked some bits and bobs, so fancy having a quick scan @rachelandrew?
> 
> I think we should add a stage in the content spreadsheet that signals that the PR is ready to rock but we just need to go through and stick final versions of diagrams in SVG format where appropriate too. That’s the sort of thing I can do in batches.

I created an issue: https://github.com/GoogleChrome/web.dev/issues/5001

If you could add any that you know need doing in there we can then check them off as they are replaced.

**Comment by una** [comment]
Ok great, as long as we're tracking #5001

**Review by una — APPROVED** [review]

**Comment by una** [comment]
Any reason to keep this as WIP?

**Comment by rachelandrew** [comment]
no we can merge these - I had flagged them before I realised they wouldn't go live :)

**Pull request merged**